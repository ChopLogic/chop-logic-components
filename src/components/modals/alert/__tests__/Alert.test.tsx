import { ChopLogicAlertMode } from '@enums';
import { ChopLogicAlertProps } from '@models';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { describe, expect, it, vi } from 'vitest';

import ChopLogicAlert from '../Alert';
import ChopLogicAlertHeader from '../elements/AlertHeader.tsx';

import { DARK_THEME } from '@css';

const defaultProps: ChopLogicAlertProps = {
  isOpened: false,
  onClose: vi.fn(),
  message: 'This is a test message.',
  title: 'Test Alert Title',
  mode: ChopLogicAlertMode.Help,
  className: 'test-class',
  id: 'test-id',
};

const renderAlert = (props: Partial<ChopLogicAlertProps> = {}) => {
  return render(<ChopLogicAlert {...defaultProps} {...props} />);
};

describe('ChopLogicAlert', () => {
  it('should render nothing when not opened', () => {
    renderAlert();
    expect(screen.queryByText(defaultProps.message)).toBeNull();
  });

  it('should render the alert when opened', () => {
    renderAlert({ isOpened: true });
    expect(screen.getByText(defaultProps.message)).toBeInTheDocument();
    expect(screen.getByText('Test Alert Title')).toBeInTheDocument();
  });

  it('should call onClose when the close button is clicked', async () => {
    const onCloseMock = vi.fn();
    renderAlert({ isOpened: true, onClose: onCloseMock });
    const closeButton = screen.getByRole('button', { name: /close/i });
    await userEvent.click(closeButton);
    expect(onCloseMock).toHaveBeenCalled();
  });

  it('should unmount the alert when it is closed', async () => {
    const { rerender } = renderAlert({ isOpened: true });
    rerender(<ChopLogicAlert {...defaultProps} isOpened={false} />);
    await waitFor(() => {
      expect(screen.queryByText('This is a test message')).not.toBeInTheDocument();
    });
  });

  it('should have different headers depending on the mode', () => {
    const { asFragment: errorFragment } = render(
      <ChopLogicAlertHeader {...defaultProps} mode={ChopLogicAlertMode.Error} theme={DARK_THEME} />,
    );
    const { asFragment: warningFragment } = render(
      <ChopLogicAlertHeader {...defaultProps} mode={ChopLogicAlertMode.Warning} theme={DARK_THEME} />,
    );
    const { asFragment: helpFragment } = render(
      <ChopLogicAlertHeader {...defaultProps} mode={ChopLogicAlertMode.Help} theme={DARK_THEME} />,
    );
    const { asFragment: infoFragment } = render(
      <ChopLogicAlertHeader {...defaultProps} mode={ChopLogicAlertMode.Info} theme={DARK_THEME} />,
    );
    const { asFragment: successFragment } = render(
      <ChopLogicAlertHeader {...defaultProps} mode={ChopLogicAlertMode.Success} theme={DARK_THEME} />,
    );

    expect(errorFragment()).toMatchSnapshot();
    expect(warningFragment()).toMatchSnapshot();
    expect(helpFragment()).toMatchSnapshot();
    expect(infoFragment()).toMatchSnapshot();
    expect(successFragment()).toMatchSnapshot();
  });
});
