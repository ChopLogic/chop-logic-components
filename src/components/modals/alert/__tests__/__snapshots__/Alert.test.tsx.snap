// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ChopLogicAlert > should have different headers depending on the mode 1`] = `
<DocumentFragment>
  <header
    class="sc-kZrBCu hFXQof"
  >
    <h3>
      <svg
        fill="currentColor"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M11 7h2v7h-2zm0 8h2v2h-2z"
        />
        <path
          d="M21.707 7.293l-5-5A.996.996 0 0016 2H8a.996.996 0 00-.707.293l-5 5A.996.996 0 002 8v8c0 .266.105.52.293.707l5 5A.996.996 0 008 22h8c.266 0 .52-.105.707-.293l5-5A.996.996 0 0022 16V8a.996.996 0 00-.293-.707zM20 15.586L15.586 20H8.414L4 15.586V8.414L8.414 4h7.172L20 8.414v7.172z"
        />
      </svg>
      Test Alert Title
    </h3>
    <button
      aria-label="Close alert popup"
      class="sc-ivxoEo dIkLYl"
      type="button"
    >
      <svg
        fill="currentColor"
        height="1em"
        viewBox="0 0 470 1000"
        width="1em"
      >
        <path
          d="M452 656c12 12 18 26.333 18 43s-6 31-18 43c-12 10.667-26.333 16-43 16s-31-5.333-43-16L234 590 102 742c-12 10.667-26.333 16-43 16s-31-5.333-43-16C5.333 730 0 715.667 0 699s5.333-31 16-43l138-156L16 342C5.333 330 0 315.667 0 299s5.333-31 16-43c12-10.667 26.333-16 43-16s31 5.333 43 16l132 152 132-152c12-10.667 26.333-16 43-16s31 5.333 43 16c12 12 18 26.333 18 43s-6 31-18 43L314 500l138 156"
        />
      </svg>
    </button>
  </header>
</DocumentFragment>
`;

exports[`ChopLogicAlert > should have different headers depending on the mode 2`] = `
<DocumentFragment>
  <header
    class="sc-kZrBCu hFXQof"
  >
    <h3>
      <svg
        fill="currentColor"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M4 20v-6a8 8 0 1116 0v6h1v2H3v-2h1zm2 0h12v-6a6 6 0 10-12 0v6zm5-18h2v3h-2V2zm8.778 2.808l1.414 1.414-2.12 2.121-1.415-1.414 2.121-2.121zM2.808 6.222l1.414-1.414 2.121 2.12L4.93 8.344 2.808 6.222zM7 14a5 5 0 015-5v2a3 3 0 00-3 3H7z"
        />
      </svg>
      Test Alert Title
    </h3>
    <button
      aria-label="Close alert popup"
      class="sc-ivxoEo dIkLYl"
      type="button"
    >
      <svg
        fill="currentColor"
        height="1em"
        viewBox="0 0 470 1000"
        width="1em"
      >
        <path
          d="M452 656c12 12 18 26.333 18 43s-6 31-18 43c-12 10.667-26.333 16-43 16s-31-5.333-43-16L234 590 102 742c-12 10.667-26.333 16-43 16s-31-5.333-43-16C5.333 730 0 715.667 0 699s5.333-31 16-43l138-156L16 342C5.333 330 0 315.667 0 299s5.333-31 16-43c12-10.667 26.333-16 43-16s31 5.333 43 16l132 152 132-152c12-10.667 26.333-16 43-16s31 5.333 43 16c12 12 18 26.333 18 43s-6 31-18 43L314 500l138 156"
        />
      </svg>
    </button>
  </header>
</DocumentFragment>
`;

exports[`ChopLogicAlert > should have different headers depending on the mode 3`] = `
<DocumentFragment>
  <header
    class="sc-kZrBCu hFXQof"
  >
    <h3>
      <svg
        fill="currentColor"
        height="1em"
        viewBox="0 0 512 512"
        width="1em"
      >
        <path
          d="M414.39 97.61A224 224 0 1097.61 414.39 224 224 0 10414.39 97.61zM192.13 260.18a64 64 0 1159.69 59.69 64.07 64.07 0 01-59.69-59.69zm240-66.64l-96.37 5.84a4.06 4.06 0 01-3.44-1.59 96 96 0 00-18.07-18.07 4.06 4.06 0 01-1.59-3.44l5.84-96.37a4 4 0 015.42-3.51A193 193 0 01435.6 188.12a4 4 0 01-3.51 5.42zM193.54 79.91l5.84 96.37a4.06 4.06 0 01-1.59 3.44 96 96 0 00-18.07 18.07 4.06 4.06 0 01-3.44 1.59l-96.37-5.84a4 4 0 01-3.51-5.42A193 193 0 01188.12 76.4a4 4 0 015.42 3.51zM79.91 318.46l96.37-5.84a4.06 4.06 0 013.44 1.59 96 96 0 0018.07 18.07 4.06 4.06 0 011.59 3.44l-5.84 96.37a4 4 0 01-5.42 3.51A193 193 0 0176.4 323.88a4 4 0 013.51-5.42zm238.55 113.63l-5.84-96.37a4.06 4.06 0 011.59-3.44 96 96 0 0018.07-18.07 4.06 4.06 0 013.44-1.59l96.37 5.84a4 4 0 013.51 5.42A193 193 0 01323.88 435.6a4 4 0 01-5.42-3.51z"
        />
      </svg>
      Test Alert Title
    </h3>
    <button
      aria-label="Close alert popup"
      class="sc-ivxoEo dIkLYl"
      type="button"
    >
      <svg
        fill="currentColor"
        height="1em"
        viewBox="0 0 470 1000"
        width="1em"
      >
        <path
          d="M452 656c12 12 18 26.333 18 43s-6 31-18 43c-12 10.667-26.333 16-43 16s-31-5.333-43-16L234 590 102 742c-12 10.667-26.333 16-43 16s-31-5.333-43-16C5.333 730 0 715.667 0 699s5.333-31 16-43l138-156L16 342C5.333 330 0 315.667 0 299s5.333-31 16-43c12-10.667 26.333-16 43-16s31 5.333 43 16l132 152 132-152c12-10.667 26.333-16 43-16s31 5.333 43 16c12 12 18 26.333 18 43s-6 31-18 43L314 500l138 156"
        />
      </svg>
    </button>
  </header>
</DocumentFragment>
`;

exports[`ChopLogicAlert > should have different headers depending on the mode 4`] = `
<DocumentFragment>
  <header
    class="sc-kZrBCu hFXQof"
  >
    <h3>
      <svg
        fill="currentColor"
        height="1em"
        viewBox="0 0 1024 1024"
        width="1em"
      >
        <path
          d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
        />
        <path
          d="M464 336a48 48 0 1096 0 48 48 0 10-96 0zm72 112h-48c-4.4 0-8 3.6-8 8v272c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V456c0-4.4-3.6-8-8-8z"
        />
      </svg>
      Test Alert Title
    </h3>
    <button
      aria-label="Close alert popup"
      class="sc-ivxoEo dIkLYl"
      type="button"
    >
      <svg
        fill="currentColor"
        height="1em"
        viewBox="0 0 470 1000"
        width="1em"
      >
        <path
          d="M452 656c12 12 18 26.333 18 43s-6 31-18 43c-12 10.667-26.333 16-43 16s-31-5.333-43-16L234 590 102 742c-12 10.667-26.333 16-43 16s-31-5.333-43-16C5.333 730 0 715.667 0 699s5.333-31 16-43l138-156L16 342C5.333 330 0 315.667 0 299s5.333-31 16-43c12-10.667 26.333-16 43-16s31 5.333 43 16l132 152 132-152c12-10.667 26.333-16 43-16s31 5.333 43 16c12 12 18 26.333 18 43s-6 31-18 43L314 500l138 156"
        />
      </svg>
    </button>
  </header>
</DocumentFragment>
`;

exports[`ChopLogicAlert > should have different headers depending on the mode 5`] = `
<DocumentFragment>
  <header
    class="sc-kZrBCu hFXQof"
  >
    <h3>
      <svg
        fill="currentColor"
        height="1em"
        viewBox="0 0 512 512"
        width="1em"
      >
        <path
          d="M470.6 105.4c12.5 12.5 12.5 32.8 0 45.3l-256 256c-12.5 12.5-32.8 12.5-45.3 0l-128-128c-12.5-12.5-12.5-32.8 0-45.3s32.8-12.5 45.3 0L192 338.7l233.4-233.3c12.5-12.5 32.8-12.5 45.3 0z"
        />
      </svg>
      Test Alert Title
    </h3>
    <button
      aria-label="Close alert popup"
      class="sc-ivxoEo dIkLYl"
      type="button"
    >
      <svg
        fill="currentColor"
        height="1em"
        viewBox="0 0 470 1000"
        width="1em"
      >
        <path
          d="M452 656c12 12 18 26.333 18 43s-6 31-18 43c-12 10.667-26.333 16-43 16s-31-5.333-43-16L234 590 102 742c-12 10.667-26.333 16-43 16s-31-5.333-43-16C5.333 730 0 715.667 0 699s5.333-31 16-43l138-156L16 342C5.333 330 0 315.667 0 299s5.333-31 16-43c12-10.667 26.333-16 43-16s31 5.333 43 16l132 152 132-152c12-10.667 26.333-16 43-16s31 5.333 43 16c12 12 18 26.333 18 43s-6 31-18 43L314 500l138 156"
        />
      </svg>
    </button>
  </header>
</DocumentFragment>
`;
